### ISSUE TITLE
Decouple NWheels.Platform.Messaging from Kestrel

### ISSUE THREAD

**Issue by felix-b** [felix-b/NWheels#29]
Just realized, we should have put Kestrel stack in NWheels.Platform.Messaging.Adapters.Kestrel. 

Additional adapters for NWheels.Platform.Messaging would include:

- NWheels.Platform.Messaging.Adapters.Wcf
- NWheels.Platform.Messaging.Adapters.Msmq
- NWheels.Platform.Messaging.Adapters.Tcp

**Comment by ILyaIL83** [comment]
Bingo!

**Comment by felix-b** [comment]
An example of robust setup for IWebHost:  https://github.com/aspnet/KestrelHttpServer/blob/dev/samples/SampleApp/Startup.cs

**Comment by felix-b** [comment]
### Ports and Adapters pattern in feature loaders

In feature loaders, a common DI pattern needs to be established for [Ports and Adapters](http://alistair.cockburn.us/Hexagonal+architecture).

Current issue is a concrete example of port-and-adapter decoupling.

### Problems and suggestions

An application-level feature loader wants to add REST API service accessible with HTTP.
- **Problem 1**: `RestApiService` component does not expose itself to the outside world. But it can handle requests received from _endpoints_.

#### Solution to Problem 1

Feature loader wants to add an HTTP endpoint, and route incoming messages to the REST API service.
- **Problem 2**: Feature loader should be decoupled from any concrete technology stack (NancyFx? ASP.NET? Kestrel? IIS integration?)

#### Solution to Problem 2

- Feature loader registers a _Port_ for an HTTP endpoint, namely, an `HttpEndpointPort` object.
- In the `HttpEndpointPort`, it describes required configuration, e.g. TCP port numbers, SSL certificates, location of static resource files, etc. And the last but not least, it contains a delegate that connects HTTP endpoint with `RestApiService`.
- A concrete HTTP stack feature loader (which must be included in the microservice), will ask DI container for all registered `HttpEndpointPort` objects (the _Ports_), and introduce an endpoint component (the _Adapter_) per each `HttpEndpointPort`.
- Problem 3: how the concrete stack feature loader is supposed to do that? First, it has to get a container and invoke `ResolveAll<HttpEndpointPort>()`. Second, it has to register endpoint components in a container builder. Technically, we have `ContributeCompiledComponents` which is passed exactly those parameters, but it's not intended for this kind of thing. Moreover, if a newly introduced endpoint component needs any component compilation, it is already too late in the lifecycle.

#### Solution to Problem 3

Introduce a new lifecycle step in `IFeatureLoader`, implemented by a new method `ContributeAdapterComponents`, with the same signatule as `ContributeCompiledComponents`:

```c#
void ContributeAdapterComponents(IComponentContainer input, IComponentContainerBuilder output)
```

The new lifecycle will be as follows:
1. ContributeConfigSections
1. ContributeComponents
1. **ContributeAdapterComponents**
1. CompileComponents
1. ContributeCompiledComponents

Any _ports_ registered in `ContributeComponents` will be backed by _adapters_ in `ContributeAdapterComponents`.

**Comment by ILyaIL83** [comment]
I don't see the reason why not to move 
`IEnumerable<Type> GetAllServices(Type baseType);`
method from IInternalComponentContainer to IComponentContainerBuilder and make IInternalComponentContainer really internal(nowadays only MicroserviceHost using it and I don't see who else can), then IAdapter can get all IAdaptable that it need.

But why IAdapter can't get all IAdaptable injected in constructor(with compiling we modify container by generating new types at all, but here we don't have, we only want to link IAdapter with IAdaptable).

**Comment by felix-b** [comment]
Regarding `GetAllServices`, it's already done in #30, except that I didn't make it `internal`. I'd prefer moving it from NWheels.Api to NWheels.Implementation instead. In this way it becomes "internal" de facto. I would avoid actually making it `internal`, because I'm afraid of blocking advanced scenarios our users may want to implement.

Regarding IAdaptable/IAdapter, this is more or less what I'm going to do. I'd like to define a clear convention. I'd like to stick to _port-and-adapter_ terminology because it's widely accepted. 

I'm currently playing in code with the following APIs:

All ports inherit from this class - to make things clear and simple (not leave it to user's decision)
```c#
public abstract class PortInjector
{
}
```

Example - HTTP endpoint port
```c#
public class HttpEndpointPortInjector : PortInjector
{
    public HttpEndpointPortInjector(
        IHttpEndpointConfiguration configuration, 
        Func<HttpContext, Task> handler)
    {
        this.Configuration = configuration;
        this.Handler = handler;
    }
    public IHttpEndpointConfiguration Configuration { get; }
    public Func<HttpContext, Task> Handler { get; }
}
```

Code that needs an HTTP endpoint, registers a port:
```c#
// handler delegate connects endpoint to message processing component (e.g. REST API service)
containerBuilder.ContributeHttpEndpoint(config, handler); 
```

ContributeHttpEndpoint implementation:
```c#
public static void ContributeHttpEndpoint<TComponent>(
	this IComponentContainerBuilder containerBuilder, 
	IHttpEndpointConfiguration configuration,
	Func<HttpContext, Task> handler)
	where TComponent : class
{
	containerBuilder.RegisterInstance<HttpEndpointPortInjector>(new HttpEndpointPortInjector(configuration, handler));
}
```

Concrete HTTP stack feature converts relevant ports to endpoint components:
```c#
[DefaultFeatureLoader]
public class KestrelFeatureLoader : FeatureLoaderBase
{
	public override void ContributeAdapterComponents(IComponentContainer input, IComponentContainerBuilder output)
	{
		base.ContributeCompiledComponents(input, output);

		var allInjectors = input.ResolveAll<HttpEndpointPortInjector>();

		foreach (var injector in allInjectors)
		{
			//... register endpoint component - TBD...
		}
	}
}
```

**Comment by ILyaIL83** [comment]
Great about https://github.com/felix-b/NWheels/issues/30.

Two questions:
1. Why do you want to work with classes, not interfaces, it shouldn't be testable(or just for brevity here, but not in the real code)?
2. Why Adapters can't register himself in ContributeComponents method, do you think they will need compiled components that's why they should be after in separate step?

**Comment by felix-b** [comment]
1. Why PortInjector and HttpEndpointPortInjector are not interfaces? They are discriminators, for different kinds of adapters, to connect to right kind of ports. Usually they will contain some configuration of the port. They are very simple DTOs, I don't see the need for abstraction. We should introduce abstractions only where they are really necessary. It will be simpler for tests to fill a concrete DTO than an abstract one.

1. Exactly.

**Comment by ILyaIL83** [comment]
1. If it's so simple and testable then I have no problem.
2. OK then.

**Comment by felix-b** [comment]
Resolved in PR #46.
